### PR TITLE
ci(feat): Test various Python versions

### PR DESCRIPTION
**Pull request by armenzg**
This will ensure that we don't regress installing dependencies for:
* our current version
* the next version we're targetting
* the version the SDK team might use

### CODE REVIEW

File: requirements-base.txt [hunk 1]
```diff
@@ -3,7 +3,8 @@ boto3==1.10.49
 botocore==1.13.50
 celery==4.4.7
 click==7.1.2
-confluent-kafka==1.5.0
+confluent-kafka==1.5.0; python_version < '3.9'
+confluent-kafka==1.6.0; python_version == '3.9'
```
Comment (armenzg) [inline requirements-base.txt:7]: This is to help with Linux on Python 3.9:
https://github.com/getsentry/sentry/pull/24394/checks?check_run_id=2097206058
```
/tmp/pip-install-d77_dxr4/confluent-kafka_62fe4d03737747d6bff6dfc1cd8ac8b4/confluent_kafka/src/confluent_kafka.h:22:10: fatal error: librdkafka/rdkafka.h: No such file or directory
       22 | #include <librdkafka/rdkafka.h>
          |          ^~~~~~~~~~~~~~~~~~~~~~
    compilation terminated.
    error: command '/usr/bin/gcc' failed with exit code 1
```

File: requirements-base.txt [hunk 2]
```diff
@@ -29,7 +30,8 @@ petname==2.6
 phonenumberslite==8.11.3
 Pillow==8.1.0
 progressbar2==3.32.0
-psycopg2-binary==2.7.7
+psycopg2-binary==2.7.7; python_version == '3.6'
+psycopg2-binary==2.8.6; python_version > '3.6'
```
Comment (armenzg) [inline requirements-base.txt:34]: This is to help with Python 3.8 & Python 3.9 on Mac and Linux:
https://github.com/getsentry/sentry/pull/24394/checks?check_run_id=2096888344
```
gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/usr/local/opt/sqlite/include -I/usr/local/opt/sqlite/include -DPSYCOPG_DEFAULT_PYDATETIME=1 -DPSYCOPG_VERSION=2.7.7 (dt dec pq3 ext lo64) -DPG_VERSION_NUM=130002 -DHAVE_LO64=1 -I/Users/runner/work/sentry/sentry/.venv/include -I/Users/runner/hostedtoolcache/Python/3.8.8/x64/include/python3.8 -I. -I/usr/local/include -I/usr/local/include/postgresql/server -c psycopg/psycopgmodule.c -o build/temp.macosx-10.14-x86_64-3.8/psycopg/psycopgmodule.o
    psycopg/psycopgmodule.c:689:18: error: incomplete definition of type 'struct _is'
        while (interp->next)
               ~~~~~~^
    /Users/runner/hostedtoolcache/Python/3.8.8/x64/include/python3.8/pystate.h:20:8: note: forward declaration of 'struct _is'
    struct _is;
           ^
    psycopg/psycopgmodule.c:690:24: error: incomplete definition of type 'struct _is'
            interp = interp->next;
                     ~~~~~~^
    /Users/runner/hostedtoolcache/Python/3.8.8/x64/include/python3.8/pystate.h:20:8: note: forward declaration of 'struct _is'
    struct _is;
           ^
    2 errors generated.
    
    It appears you are missing some prerequisite to build the package from source.
    
    You may install a binary package by installing 'psycopg2-binary' from PyPI.
    If you want to install psycopg2 from source, please install the packages
    required for the build and try again.
    
    For further information please check the 'doc/src/install.rst' file (also at
    <http://initd.org/psycopg/docs/install.html>).
    
    error: command 'gcc' failed with exit status 1
```

File: requirements-base.txt [hunk 3]
```diff
@@ -73,7 +75,8 @@ grpcio==1.35.0
 hiredis==0.3.1
 
 # not directly used, but pinned for at least semaphore/symbolic
-cffi==1.12.2
+cffi==1.12.2; python_version == '3.6'
+cffi==1.14.4; python_version > '3.6'
```
Comment (armenzg) [inline requirements-base.txt:79]: This is to help with Python 3.9 (affecting at least Linux):
https://github.com/getsentry/sentry/pull/24394/checks?check_run_id=2097113005
```
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DUSE__THREAD -DHAVE_SYNC_SYNCHRONIZE -I/home/linuxbrew/.linuxbrew/Cellar/libffi/3.3_3/include -I/home/runner/work/sentry/sentry/.venv/include -I/opt/hostedtoolcache/Python/3.9.2/x64/include/python3.9 -c c/_cffi_backend.c -o build/temp.linux-x86_64-3.9/c/_cffi_backend.o
    c/_cffi_backend.c: In function ‘prepare_callback_info_tuple’:
    c/_cffi_backend.c:5987:5: warning: ‘PyEval_InitThreads’ is deprecated [-Wdeprecated-declarations]
     5987 |     PyEval_InitThreads();
          |     ^~~~~~~~~~~~~~~~~~
    In file included from /opt/hostedtoolcache/Python/3.9.2/x64/include/python3.9/Python.h:145,
                     from c/_cffi_backend.c:2:
    /opt/hostedtoolcache/Python/3.9.2/x64/include/python3.9/ceval.h:130:37: note: declared here
      130 | Py_DEPRECATED(3.9) PyAPI_FUNC(void) PyEval_InitThreads(void);
          |                                     ^~~~~~~~~~~~~~~~~~
    c/_cffi_backend.c: In function ‘b_callback’:
    c/_cffi_backend.c:6042:5: warning: ‘ffi_prep_closure’ is deprecated: use ffi_prep_closure_loc instead [-Wdeprecated-declarations]
     6042 |     if (ffi_prep_closure(closure, &cif_descr->cif,
          |     ^~
    In file included from c/_cffi_backend.c:15:
    /home/linuxbrew/.linuxbrew/Cellar/libffi/3.3_3/include/ffi.h:334:1: note: declared here
      334 | ffi_prep_closure (ffi_closure*,
          | ^~~~~~~~~~~~~~~~
    In file included from c/cffi1_module.c:20,
                     from c/_cffi_backend.c:7552:
    c/call_python.c: In function ‘_get_interpstate_dict’:
    c/call_python.c:26:30: error: dereferencing pointer to incomplete type ‘PyInterpreterState’ {aka ‘struct _is’}
       26 |     builtins = tstate->interp->builtins;
          |                              ^~
    error: command '/usr/bin/gcc' failed with exit code 1
```

File: requirements-base.txt [hunk 4]
```diff
@@ -73,7 +76,8 @@ grpcio==1.35.0
 hiredis==0.3.1
 
 # not directly used, but pinned for at least semaphore/symbolic
-cffi==1.12.2
+cffi==1.12.2; python_version <= '3.8'
+cffi==1.14.4; python_version == '3.9'
```
Comment (armenzg) [inline requirements-base.txt:80]: This is to help with Python 3.9 (affecting at least Linux):
https://github.com/getsentry/sentry/pull/24394/checks?check_run_id=2097113005
```
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DUSE__THREAD -DHAVE_SYNC_SYNCHRONIZE -I/home/linuxbrew/.linuxbrew/Cellar/libffi/3.3_3/include -I/home/runner/work/sentry/sentry/.venv/include -I/opt/hostedtoolcache/Python/3.9.2/x64/include/python3.9 -c c/_cffi_backend.c -o build/temp.linux-x86_64-3.9/c/_cffi_backend.o
    c/_cffi_backend.c: In function ‘prepare_callback_info_tuple’:
    c/_cffi_backend.c:5987:5: warning: ‘PyEval_InitThreads’ is deprecated [-Wdeprecated-declarations]
     5987 |     PyEval_InitThreads();
          |     ^~~~~~~~~~~~~~~~~~
    In file included from /opt/hostedtoolcache/Python/3.9.2/x64/include/python3.9/Python.h:145,
                     from c/_cffi_backend.c:2:
    /opt/hostedtoolcache/Python/3.9.2/x64/include/python3.9/ceval.h:130:37: note: declared here
      130 | Py_DEPRECATED(3.9) PyAPI_FUNC(void) PyEval_InitThreads(void);
          |                                     ^~~~~~~~~~~~~~~~~~
    c/_cffi_backend.c: In function ‘b_callback’:
    c/_cffi_backend.c:6042:5: warning: ‘ffi_prep_closure’ is deprecated: use ffi_prep_closure_loc instead [-Wdeprecated-declarations]
     6042 |     if (ffi_prep_closure(closure, &cif_descr->cif,
          |     ^~
    In file included from c/_cffi_backend.c:15:
    /home/linuxbrew/.linuxbrew/Cellar/libffi/3.3_3/include/ffi.h:334:1: note: declared here
      334 | ffi_prep_closure (ffi_closure*,
          | ^~~~~~~~~~~~~~~~
    In file included from c/cffi1_module.c:20,
                     from c/_cffi_backend.c:7552:
    c/call_python.c: In function ‘_get_interpstate_dict’:
    c/call_python.c:26:30: error: dereferencing pointer to incomplete type ‘PyInterpreterState’ {aka ‘struct _is’}
       26 |     builtins = tstate->interp->builtins;
          |                              ^~
    error: command '/usr/bin/gcc' failed with exit code 1
```

File: .github/workflows/python-versions.yml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  # optionally c) a team request
+  python-versions:
+    name: set up
+    runs-on: ${{ matrix.os }}
+    timeout-minutes: 15
+    strategy:
+      matrix:
+        # macosx-10.15 is Catalina
+        # macosx-11.0 is Big Sur, however, it takes long for jobs to get started
+        os: [ macos-10.15, macos-11.0, ubuntu-20.04 ]
```
Comment (armenzg) [inline .github/workflows/python-versions.yml:20]: I've added Big Sur last moment which might yield some errors.
It's far more important for me to get your initial feedback.

**Review by armenzg — COMMENTED** [review]
This work can help a bit with the migration to 3.8 since it can show packages that need updating, however, the main purpose of this PR is not to migrate to 3.8 but to allow supporting multiple Python versions.

File: requirements-base.txt (hunk 4 above)
Comment (jan-auer) [reply requirements-base.txt:80]: What's the reason for not just upgrading to 1.14?

Comment (armenzg) [reply requirements-base.txt:80]: We could. I just didn't want to deal with upgrading it in this PR since I would have to check release notes et al.

**Comment by armenzg** [comment]
@getsentry/owners-python-build @getsentry/owners-sentry-dev Could I please have a review? Thanks!

File: requirements-base.txt (hunk 4 above)
Comment (joshuarli) [reply requirements-base.txt:80]: I'll take care of upgrading cffi, it's on my list for py38.

**Review by joshuarli — APPROVED** [review]

**Review by joshuarli — CHANGES_REQUESTED** [review]
Actually, I'm not sure about this. At the very least I'd request that we don't test 3.9; we aren't targetting it anytime soon.

File: .github/workflows/python-versions.yml [hunk 6]
```diff
@@ -0,0 +1,60 @@
+name: python versions
+on:
+  pull_request:
+    paths:
+    - 'requirements*'
+    - '.github/workflows/python-versions.yml'
```
[the remark this answers is not in the document]
Comment (armenzg) [reply .github/workflows/python-versions.yml:6]: This line triggers the workflow if changes are made to it, otherwise, it doesn't run.

**Comment by armenzg** [comment]
> Actually, I'm not sure about this. At the very least I'd request that we don't test 3.9; we aren't targetting it anytime soon.

Some folks like Jan and others that don't touch sentry often tend to use Python 3.9 because that's what comes with their host.

With a little bit of work (back in Dec/Jan), I had enabled him and others to use Python 3.9 if they wish to do so. When I hard pinned the requirements I broke that support.

If support of Python 3.9 was to require a significant effort I would not consider pushing for it.

@joshuarli what other concerns do you have? Or does my response alleviate them?

**Comment by joshuarli** [comment]
> > Actually, I'm not sure about this. At the very least I'd request that we don't test 3.9; we aren't targetting it anytime soon.
> 
> Some folks like Jan and others that don't touch sentry often tend to use Python 3.9 because that's what comes with their host.
> 
> With a little bit of work (back in Dec/Jan), I had enabled him and others to use Python 3.9 if they wish to do so. When I hard pinned the requirements I broke that support.
> 
> If support of Python 3.9 was to require a significant effort I would not consider pushing for it.
> 
> @joshuarli what other concerns do you have? Or does my response alleviate them?

We can keep the requirement pins, but I'd really like to just support people running Sentry on current or targetted versions. Otherwise we aren't solving any immediate problems. If they're running 3.9 then that's their responsibility. This comes from a place of wanting to run as little CI as possible. Even if this is just going to run on requirements change, it's still a lot of extra jobs. Requirements is going to be changed a lot going forward into py38 by me as well. I don't care about 3.9.

**Comment by billyvg** [comment]
Is there a technical reason why the SDK teams need to run the host python version?

If they can't develop using pyenv for Sentry then we should definitely have these checks in place to make sure 3.9 works for them. Otherwise, if it's not due to technical limitations, then I'm more +1 on not maintaining support.

Although in general, I am in favor of testing against future versions just so that we are not completely blind-sided by upgrades. Though in these cases, it might be better to run as a scheduled workflow.

File: .github/workflows/python-versions.yml [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    strategy:
+      matrix:
+        # macosx-10.15 is Catalina
+        # macosx-11.0 is Big Sur, however, it takes long for jobs to get started
+        os: [ macos-10.15, macos-11.0, ubuntu-20.04 ]
+        python-version: [ 3.6, 3.8, 3.9 ]
+      fail-fast: false
+    env:
+      PIP_DISABLE_PIP_VERSION_CHECK: on
+      # There's a check that prevents make install-py-dev to work if the develop has not
```
Comment (billyvg) [inline .github/workflows/python-versions.yml:25]:
```suggestion
      # There's a check that prevents make install-py-dev to work if the developer has not
```

**Comment by untitaker** [comment]
@mitsuhiko can elaborate on this but I think some SDK folks chose 3.9 around the time Big Sur was fresh and broke the world including pyenv. I'm using 3.6 just fine. I don't think we have to maintain support if it's a PITA

File: .github/workflows/python-versions.yml [hunk 8]
```diff
@@ -0,0 +1,59 @@
+name: python versions
+on:
+  schedule:
+    # Notifications will be sent to whoever touches this last
+    # You will be able to trigger this workflow manually if need be
+    - cron:  '0 0 * * 7'
```
Comment (armenzg) [inline .github/workflows/python-versions.yml:6]: Schedule once a week.

**Comment by armenzg** [comment]
I've moved this to run once a week.

File: .github/workflows/python-versions.yml [hunk 9]
```diff
@@ -0,0 +1,59 @@
+name: python versions
+on:
+  schedule:
+    # Notifications will be sent to whoever touches this last
+    # You will be able to trigger this workflow manually if need be
+    - cron:  '0 0 * * 0'
```
Comment (joshuarli) [inline .github/workflows/python-versions.yml:6]: Running once a week is nice, I think this is a good compromise vs. changing on requirements changes. My only recommendation would be to use a better offtime:

```suggestion
    # This runs at 4:00 every Saturday, where GHA should in theory be less busy.
    - cron:  '0 4 * * 6'
```

image

**Review by joshuarli — APPROVED** [review]
I like the once a week, if you wanna run 3.9 then, that's fine given that this is an infrequent cronjob. Though I still don't think it's necessary, plus we'd wanna encourage people to use pyenv. If 3.9 works then they likely just won't use pyenv.

**Pull request merged**